A15-binary-logging

```diff
@@ -0,0 +1,264 @@
+
+This functionality should be added in each of our supported languages. The ease
+of adoption should also be considered.
+
+## Proposal
+
+The proposed design is adding binary logging as a built-in functionality in
+gRPC. Users can turn it on/off by setting the `GRPC_BINARY_LOG_FILTER`
+environment variable. It's not required to modify application code to generate
+logs. The control interface should allow users to change the amount and contents
+of logs.
+
+### Log Format
+
+gRPC logs each RPC in the format defined by proto GrpcLogEntry. Initial/trailing
+metadata and messages are logged separately to match the on-the-wire
+representation.
+
+The proto of GrpcLogEntry:
+
+```
+// Log entry we store in binary logs
+message GrpcLogEntry {
+  // Enumerates the type of logs
+  enum Type {
+    SEND_INITIAL_METADATA = 0;
+    SEND_TRAILING_METADATA = 1;
+    SEND_MESSAGE = 2;
+    RECV_INITIAL_METADATA = 3;
+    RECV_TRAILING_METADATA = 4;
+    RECV_MESSAGE = 5;
+  };
+
+  // Enumerates the entity that generates the log entry
+  enum Logger {
+    CLIENT = 0;
+    SERVER = 1;
+  }
+
+  Type type = 1;      // One of the above Type enum
+  Logger logger = 2;  // One of the above Logger enum
+
+  // Uniquely identifies a call. Each call may have several log entries, they
+  // will share the same call_id. 128 bits split into 2 64-bit parts.
+  Uint128 call_id = 3;
+
+  // The logger uses one of the following fields to record the payload,
+  // according to the type of the log entry.
+  oneof payload {
+    // Used by CLIENT_INIT_METADATA, SERVER_INIT_METADATA and TRAILING_METADATA
+    Metadata metadata = 4;
+    // Used by REQUEST and RESPONSE
+    Message message = 5;
+  }
+
+  // Peer address information, will only be recorded in SEND_INITIAL_METADATA
+  // and RECV_INITIAL_METADATA entries.
+  Peer peer = 6;
+};
+
+// Message payload, used by REQUEST and RESPONSE
+message Message {
+  // This flag is currently used to indicate whether the payload is compressed,
+  // it may contain other semantics in the future. Value of 1 indicates that the
+  // binary octet sequence of Message is compressed using the mechanism declared
+  // by the Message-Encoding header. A value of 0 indicates that no encoding of
+  // Message bytes has occurred.
+  uint8 flags = 1;
+  // Length of the message. It may not be the same as the length of the
+  // data field, as the logging payload can be truncated or omitted.
+  uint32 length = 1;
+  // Binary payload, post compression (depends on the flags field), may be
+  // truncated or omitted.
+  bytes data = 2;
+}
+
+// A list of metadata pairs, used in the payload of CLIENT_INIT_METADATA,
+// SERVER_INIT_METADATA and TRAILING_METADATA
+message Metadata {
+  repeated MetadataEntry entry = 1;
+}
+
+// A metadata key value pair
+message MetadataEntry {
+  string key = 1;
+  bytes value = 2;
+}
+
+// Peer information
+message Peer {
+  enum PeerType {
+    PEER_IPV4 = 0;  // peer is struct sockaddr_in
```

Comment:
This should also have a `PEER_UNKNOWN = 0`.

Reply:
The current proto has an `UNKOWN`